packages/engine/paima-sm/src/delegate-wallet.ts

```diff
@@ -100,53 +103,70 @@ export class DelegateWallet {
     throw new Error(`Invalid Signature for ${walletAddress} : ${message}`);
   }
 
-  public async getOrCreateNewAddress(
-    address: string
-  ): Promise<{ address: IGetAddressFromAddressResult; isNew: boolean }> {
-    const [exitingAddress] = await getAddressFromAddress.run({ address: address }, this.DBConn);
-    if (exitingAddress) return { address: exitingAddress, isNew: false };
-    // create new.
+  public async createAddress(address: string): Promise<IGetAddressFromAddressResult> {
     await newAddress.run({ address: address }, this.DBConn);
     const [createdAddress] = await getAddressFromAddress.run({ address }, this.DBConn);
-    return { address: createdAddress, isNew: true };
+    return createdAddress;
+  }
+
+  public async getAddress(address: string): Promise<AddressWrapper> {
+    const [exitingAddress] = await getAddressFromAddress.run({ address: address }, this.DBConn);
+    if (!exitingAddress) return { address, exists: false };
+    return { address: exitingAddress, exists: true };
   }
 
   // Swap:
-  // addressA gains addressB ID
-  // addressB is assigned a new ID
-  private async swap(addressA: string, addressB: string): Promise<void> {
-    await deleteAddress.run({ address: addressA }, this.DBConn);
+  // addressB gains addressA ID.
+  // addressA is assigned a new ID
+  private async swap(
+    addressA: AddressExists,
+    addressB: AddressDoesNotExist
+  ): Promise<{
+    newA: AddressExists;
+    newB: AddressExists;
+  }> {
     await updateAddress.run(
       {
-        new_address: addressA,
-        old_address: addressB,
+        new_address: addressB.address,
+        old_address: addressA.address.address,
       },
       this.DBConn
     );
-    await newAddress.run({ address: addressB }, this.DBConn);
+    await newAddress.run({ address: addressA.address.address }, this.DBConn);
+
+    return {
+      newA: (await this.getAddress(addressA.address.address)) as AddressExists,
+      newB: (await this.getAddress(addressB.address)) as AddressExists,
+    };
```

Comment:
prefer `Promise.all` if we really can't do a single query

Reply:
Queries merged